### PR TITLE
Change bahavior of table upcast to read width and height from figure …

### LINKED ISSUE #12812: With GHS disabled table's styles are moved to figure's styles
## 📝 Provide detailed reproduction steps (if any)

1. With GHS disabled, enter Source mode and paste the following code:
```html
<figure class="table" style="width:75%">
    <table style="width:100%">
        <tbody>
            <tr>
                <td>
                    A cell
                </td>
            </tr>
        </tbody>
    </table>
</figure>
```

2. Exit Source mode, the content is parsed to:
Screenshot 2022-11-07 at 16 28 18

And its source is:

Screenshot 2022-11-07 at 16 28 25

### ✔️ Expected result

The figure's width should be set to `75%`.

### ❌ Actual result

The figure's width is set to `100%`.

## 📃 Other details

This issue blocks a fix for a IFW bug.

### PR DESCRIPTION
**Pull request by mremiszewski**
…element.

### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Fix (table): On upcast, skip width and height on table element if the figure element exists. Closes #12812.

---

### Additional information

_For example – encountered issues, assumptions you had to make, other affected tickets, etc._

### CODE REVIEW

File: packages/ckeditor5-table/src/converters/tableproperties.js [hunk 1]
```diff
@@ -19,7 +19,7 @@
  * @param {Boolean} [options.reduceBoxSides=false]
  */
 export function upcastStyleToAttribute( conversion, options ) {
-	const { viewElement, defaultValue, modelAttribute, styleName, reduceBoxSides = false } = options;
+	const { viewElement, defaultValue, modelAttribute, styleName, reduceBoxSides = false, skipCondition = () => false } = options;
```
Comment (niegowski) [inline packages/ckeditor5-table/src/converters/tableproperties.js:22]: Maybe we could find a better name? Also, remember to add this option to the JsDoc above.

File: packages/ckeditor5-table/src/tableproperties/tablepropertiesediting.js [hunk 2]
```diff
@@ -93,14 +93,16 @@ export default class TablePropertiesEditing extends Plugin {
...
+			skipCondition: element => element.name == 'table' && element.parent.name == 'figure'
 		} );
 		editor.commands.add( 'tableWidth', new TableWidthCommand( editor, defaultTableProperties.width ) );
 
 		enableTableToFigureProperty( schema, conversion, {
 			modelAttribute: 'tableHeight',
 			styleName: 'height',
-			defaultValue: defaultTableProperties.height
+			defaultValue: defaultTableProperties.height,
+			skipCondition: element => element.name == 'table' && element.parent.name == 'figure'
```
Comment (niegowski) [inline packages/ckeditor5-table/src/tableproperties/tablepropertiesediting.js:105]: `enableTableToFigureProperty()` is a dedicated helper that should handle `figure` elements so this option should be passed inside this helper.

File: packages/ckeditor5-table/tests/tableproperties/tablepropertiesediting.js [hunk 3]
```diff
@@ -957,6 +957,36 @@ describe( 'table properties', () => {
 
 					expect( table.getAttribute( 'tableWidth' ) ).to.equal( '1337px' );
 				} );
+
+				it( 'should upcast width from <figure> if both <figure> and <table> has width style set', () => {
```
Comment (niegowski) [inline packages/ckeditor5-table/tests/tableproperties/tablepropertiesediting.js:961]: Please add tests for height also.

File: packages/ckeditor5-table/tests/tableproperties/tablepropertiesediting.js [hunk 4]
```diff
@@ -957,6 +957,36 @@ describe( 'table properties', () => {
...
+					editor.setData(
+						'<figure class="table" style="width:75%">' +
+							'<table style="width:95%"><tbody><tr><td>1</td></tr></tbody></table>' +
+						'</figure>'
+					);
```
Comment (niegowski) [inline packages/ckeditor5-table/tests/tableproperties/tablepropertiesediting.js:966]:
```suggestion
					);
```

File: packages/ckeditor5-table/tests/tableproperties/tablepropertiesediting.js [hunk 5]
```diff
@@ -957,6 +957,36 @@ describe( 'table properties', () => {
...
+					const expectModel = data => expect( getModelData( model, { withoutSelection: true } ) ).to.equalMarkup( data );
+
+					expectModel(
+						'<table tableWidth="75%">' +
+						'<tableRow><tableCell><paragraph>1</paragraph></tableCell></tableRow>' +
+						'</table>'
```
Comment (niegowski) [inline packages/ckeditor5-table/tests/tableproperties/tablepropertiesediting.js:972]:
```suggestion
						'<table tableWidth="75%">' +
							'<tableRow><tableCell><paragraph>1</paragraph></tableCell></tableRow>' +
						'</table>'
```

File: packages/ckeditor5-table/tests/tableproperties/tablepropertiesediting.js [hunk 6]
```diff
@@ -957,6 +957,36 @@ describe( 'table properties', () => {
...
 					expect( table.getAttribute( 'tableWidth' ) ).to.equal( '1337px' );
 				} );
+
+				it( 'should upcast width from <figure> if both <figure> and <table> has width style set', () => {
+					editor.setData(
+						'<figure class="table" style="width:75%">' +
+							'<table style="width:95%"><tbody><tr><td>1</td></tr></tbody></table>' +
+						'</figure>'
+					);
+					const expectModel = data => expect( getModelData( model, { withoutSelection: true } ) ).to.equalMarkup( data );
```
Comment (niegowski) [inline packages/ckeditor5-table/tests/tableproperties/tablepropertiesediting.js:967]: Let's simplify this.

File: packages/ckeditor5-table/src/converters/tableproperties.js [hunk 7]
```diff
@@ -17,9 +17,10 @@
  * @param {String} options.viewElement The view element name that should be converted.
  * @param {String} options.defaultValue The default value for the specified `modelAttribute`.
  * @param {Boolean} [options.reduceBoxSides=false]
+ * @param {Function} [options.isApllicable] The function which returns `true` if style should be checked for this element.
  */
 export function upcastStyleToAttribute( conversion, options ) {
-	const { viewElement, defaultValue, modelAttribute, styleName, reduceBoxSides = false } = options;
+	const { viewElement, defaultValue, modelAttribute, styleName, reduceBoxSides = false, isApplicable = () => true } = options;
```
Comment (niegowski) [inline packages/ckeditor5-table/src/converters/tableproperties.js:23]: Maybe `checkElement` would be a better name?

File: packages/ckeditor5-table/src/tableproperties/tablepropertiesediting.js [hunk 8]
```diff
@@ -242,6 +242,12 @@ function enableTableToFigureProperty( schema, conversion, options ) {
 	schema.extend( 'table', {
 		allowAttributes: [ modelAttribute ]
 	} );
-	upcastStyleToAttribute( conversion, { viewElement: /^(table|figure)$/, ...options } );
+
+	upcastStyleToAttribute( conversion, {
+		viewElement: /^(table|figure)$/,
+		isApplicable: element => element.name != 'table' || element.parent.name != 'figure',
```
Comment (niegowski) [inline packages/ckeditor5-table/src/tableproperties/tablepropertiesediting.js:248]:
```suggestion
		isApplicable: element => !( element.name == 'table' && element.parent.name == 'figure' ),
```

this way it's more readable IMHO.

File: packages/ckeditor5-table/src/converters/tableproperties.js (hunk 7 above)
Comment (mremiszewski) [reply packages/ckeditor5-table/src/converters/tableproperties.js:23]: I think that `checkElement` tells less thank current name.

**Review by niegowski — APPROVED** [review]

**Pull request merged**